exec: add templating for RANK and ROW_NUMBER window functions

yuzefovich opened the pull request:
Templates out rankOp into two operators (for dense and non-dense
case) and templates out support for PARTITION BY clause for both
rankOp and rowNumberOp. The code is undertested and is lacking
benchmarks, but that will be addressed soon.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/37282)
<!-- Reviewable:end -->

jordanlewis reviewed (commented):
I think you should add some lightweight benchmarks to these. Also add the .eg.go files to .gitignore.

> Reviewed 5 of 6 files at r1, 5 of 5 files at r2, 5 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37282#-:-LeFJToTBDdkq-Js_uoA:bv0umr0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, @solongordon, and @yuzefovich)

___
*[pkg/sql/exec/vecbuiltins/rank_tmpl.go, line 52 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/37282#-LeFPJFgD7O2e1UDOWbF:-LeFPJFgD7O2e1UDOWbG:b-9erwb7) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae649523e6da9f2ec5d18a953abcadea224fb60b/pkg/sql/exec/vecbuiltins/rank_tmpl.go#L52)):*
> ```Go
> // {{range .}}
> 
> type rankDense__DENSE_HasPartition__PARTITION_Op struct {
> ```

You could make an embedded struct that has all of the data that's repeated in these, to avoid the nasty initialization above.
___
*[pkg/sql/exec/vecbuiltins/rank_tmpl.go, line 103 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/37282#-LeFPaJbEG-1wpVY-grU:-LeFPaJc3FvBsSPymNHG:b-jrkbaq) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae649523e6da9f2ec5d18a953abcadea224fb60b/pkg/sql/exec/vecbuiltins/rank_tmpl.go#L103)):*
> ```Go
> 	rankCol := r.batch.ColVec(r.outputColIdx).Int64()
> 	sel := r.batch.Selection()
> 	if sel != nil {
> ```

I would probably try to template sel vs nonsel here as well since there's so much code in the body.
___
*[pkg/sql/exec/vecbuiltins/row_number.go, line 59 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37282#-LeFMcyV89t2zJnXAPfm:-LeFMcyV89t2zJnXAPfn:bar6c7y) ([raw file](https://github.com/cockroachdb/cockroach/blob/a10b4e69fee8fa3855a29bcb488d5098a2fd0872/pkg/sql/exec/vecbuiltins/row_number.go#L59)):*
> ```Go
> 		// TODO(yuzefovich): I couldn't figure out how to pass the batch as an
> 		// argument, so I embedded it into the struct. Is it possible to pass it as
> 		// an argument?
> ```

You can pass it as an argument to this function, yes. The template "function" can see everything in its surrounding scope - so what I usually do is give the template function an argument of the same name as the equivalent in the outer scope, so it looks like ordinary Go code.
___
*[pkg/sql/exec/vecbuiltins/row_number_tmpl.go, line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37282#-LeFMsEtB-EtPACMlmEh:-LeFMsEtB-EtPACMlmEi:bdpr1hw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a10b4e69fee8fa3855a29bcb488d5098a2fd0872/pkg/sql/exec/vecbuiltins/row_number_tmpl.go#L29)):*
> ```Go
> 
> // {{/*
> func _NEXT_ROW_NUMBER(hasPartition bool) { // */}}
> ```

So I'd give this guy a `batch coldata.Batch` argument, and also give `nextBodyWithPartition` that same named argument, and things should just work. 


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
I left TODOs for the benchmarks and for sel vs non-sel cases in rankOp. I want to get it in as is (undertested with no benchmarks) since it's already a pretty big PR.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37282#-:-Lf29oiBE_MPMrVIwaRR:btbaxil)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @solongordon)

___
*[pkg/sql/exec/vecbuiltins/rank_tmpl.go, line 52 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/37282#-LeFPJFgD7O2e1UDOWbF:-Lf26wU00b09kpB8hetU:bknht0j) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae649523e6da9f2ec5d18a953abcadea224fb60b/pkg/sql/exec/vecbuiltins/rank_tmpl.go#L52)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

You could make an embedded struct that has all of the data that's repeated in these, to avoid the nasty initialization above.
</blockquote></details>

Looks better, thanks!
___
*[pkg/sql/exec/vecbuiltins/rank_tmpl.go, line 103 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/37282#-LeFPaJbEG-1wpVY-grU:-Lf2ATxB-vZOWYE10ThL:b-aemzb1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae649523e6da9f2ec5d18a953abcadea224fb60b/pkg/sql/exec/vecbuiltins/rank_tmpl.go#L103)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I would probably try to template sel vs nonsel here as well since there's so much code in the body.
</blockquote></details>

Also left a TODO for this.
___
*[pkg/sql/exec/vecbuiltins/row_number.go, line 59 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37282#-LeFMcyV89t2zJnXAPfm:-Lf25ih-A_uC61sbYTb8:b-ca64jr) ([raw file](https://github.com/cockroachdb/cockroach/blob/a10b4e69fee8fa3855a29bcb488d5098a2fd0872/pkg/sql/exec/vecbuiltins/row_number.go#L59)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

You can pass it as an argument to this function, yes. The template "function" can see everything in its surrounding scope - so what I usually do is give the template function an argument of the same name as the equivalent in the outer scope, so it looks like ordinary Go code.
</blockquote></details>

Indeed, that worked, thanks!
___
*[pkg/sql/exec/vecbuiltins/row_number_tmpl.go, line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37282#-LeFMsEtB-EtPACMlmEh:-Lf25kze3XtcmfaR0NSH:b-dtkds3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a10b4e69fee8fa3855a29bcb488d5098a2fd0872/pkg/sql/exec/vecbuiltins/row_number_tmpl.go#L29)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

So I'd give this guy a `batch coldata.Batch` argument, and also give `nextBodyWithPartition` that same named argument, and things should just work. 
</blockquote></details>

Done, thanks!


<!-- Sent from Reviewable.io -->

yuzefovich commented:
RFAL guys.

jordanlewis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37282#-:-LfNV2Rq1evf4SyUf_9I:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, @solongordon, and @yuzefovich)

___
*[pkg/sql/exec/execgen/cmd/execgen/rank_gen.go, line 31 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/37282#-LfNUtQU0oUjcvVFcEgV:-LfNUtQV1P0jIcl-Qamc:bwncudn) ([raw file](https://github.com/cockroachdb/cockroach/blob/48d7011bb14e19bba1dc6ecb698dfd22f966804d/pkg/sql/exec/execgen/cmd/execgen/rank_gen.go#L31)):*
> ```Go
> }
> 
> func (r rankTmplInfo) UpdateRank() string {
> ```

Comments on these would be nice. In general I worry that this code will be confusing as it is far removed from where it is used, but that's kind of a limitation of the model in general. I think comments are the best we can do for now.


<!-- Sent from Reviewable.io -->

yuzefovich committed:
exec: template out RANK and ROW_NUMBER

Templates out rankOp into two operators (for dense and non-dense
case) and templates out support for PARTITION BY clause for both
rankOp and rowNumberOp. The code is undertested and is lacking
benchmarks, but that will be addressed soon.

Release note: None

yuzefovich commented:
TFTR!

bors r+
